pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -223,35 +253,57 @@ func (tb *TrustBundle) AddMeshConfigUpdate(cfg *meshconfig.MeshConfig) error {
 			trustBundleLog.Errorf("failed to update meshConfig PEM trustAnchors: %v", err)
 			return err
 		}
-
-		tb.updateRemoteEndpoint(endpoints)
+		tb.endpointMutex.Lock()
+		tb.endpointCfg = endpoints
+		tb.endpointMutex.Unlock()
+		tb.endpointUpdateChan <- struct{}{}
 	}
 	return nil
 }
 
-func (tb *TrustBundle) fetchRemoteTrustAnchors() {
+func (tb *TrustBundle) updateRemoteTrustAnchors() {
 	var err error
+	var remoteCerts []string
+	var fetchFail bool = false
 
 	tb.endpointMutex.RLock()
-	remoteEndpoints := tb.endpoints
+	endpointCfg := make([]string, len(tb.endpointCfg))
+	copy(endpointCfg, tb.endpointCfg)
 	tb.endpointMutex.RUnlock()
-	remoteCerts := []string{}
 
 	currentTrustDomain := spiffe.GetTrustDomain()
-	for _, endpoint := range remoteEndpoints {
+	tb.updateCfgRemoteEndpoint(endpointCfg)
+	remoteEndpoints := tb.endpoints
+
+	for endpointURI, remoteTrustAnchor := range remoteEndpoints {
 		trustDomainAnchorMap, err := spiffe.RetrieveSpiffeBundleRootCerts(
-			map[string]string{currentTrustDomain: endpoint}, tb.remoteCaCertPool, remoteTimeout)
-		if err != nil {
-			trustBundleLog.Errorf("unable to fetch trust Anchors from endpoint %s: %s", endpoint, err)
-			continue
-		}
-		certs := trustDomainAnchorMap[currentTrustDomain]
-		for _, cert := range certs {
-			certStr := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
-			trustBundleLog.Debugf("from endpoint %v, fetched trust anchor cert: %v", endpoint, certStr)
-			remoteCerts = append(remoteCerts, certStr)
+			map[string]string{currentTrustDomain: endpointURI}, tb.remoteCaCertPool, remoteTimeout)
+		if err == nil {
+			remoteTrustAnchor.Reset()
+			certs := trustDomainAnchorMap[currentTrustDomain]
+			for _, cert := range certs {
+				certStr := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
+				trustBundleLog.Debugf("from endpoint %v, fetched trust anchor cert: %v", endpointURI, certStr)
+				remoteCerts = append(remoteCerts, certStr)
+				remoteTrustAnchor.AddCert(certStr)
+			}
+		} else {
+			fetchFail = true
+			trustBundleLog.Infof("unable to fetch trustAnchor from endpoint %s: %v", endpointURI, err)
+			if !remoteEndpoints[endpointURI].IncrementRetryOrFail() {
+				remoteTrustAnchor.Reset()
```

Comment:
@myidpt - would like your input here as well. @howardjohn suggests that we shouldn't purge trustAnchors from remoteEndpoints even if they are unavailable for an entire day. My concern is that we need to have an upper limit on the time duration that these remote trustAnchors persist in the mesh. Any comments?

Reply:
Just to clarify a bit more,  I am a bit concerned about nearly ANY action that is only triggered after a day. It seems like a ticking time bomb that will not be exercised in 99.999% of cases then give someone a nasty mesh wide outage. It reminds me a bit of the "1 year Istio birthday gift" when the root cert expires after 1 year in the past (now its 10yr by default).

Reply:
Fair, and the time duration can be changed. However, note that this won't happen unless the remote Endpoint becomes unreachable. Would it perhaps make sense to raise error logs when endpoints become unreachable, alerting them to the possibility that mesh Traffic will be disrupted in the foreseeable future?

Reply:
We shouldn't remove the cached root if the remote endpoint is unreachable, because endpoint unreachable is an availability issue, not a change in security config. Rasing error logs makes sense. The root should be removed only if the user removes the endpoint from the MeshConfig.

If istiod is restarted and the cached root is lost, I think that's fine. It's due to the unavailability of the remote endpoint.

We can point this out in the documentation.